### PR TITLE
feat: 2588 add als

### PR DESCRIPTION
**Pull request by tdaly61**
Hi folks, 
I think it is now time to get a lot more eyes on this v13 -v14 upgrade work.
The account-lookup-service has subcharts and has been my learning project for how to upgrade the ML charts from V13 -V14 in the example structure @mdebarros has established.  

Can I please get your eyes on this an in particular with attention to the following:-

- 	parent HELM chart account-lookup-service is populated via the ./scripts/populate_values.py script this script copies the longest globals section from the subcharts and then copies the entire subcharts  values.yaml to the parent chart (see account-lookup-service/values.yaml )
- Please check the subcharts values.yaml files carefully and provide feedback 
- Please inspect closely the subcharts templates/deployment.yaml
- what in the heck to set pod and node affinities to by default in the values.yaml files ? 
- Please note the plan is to update the chart naming to better reflect the reference architecture after most of the updates are done for v14
- should  we create an  example_backend _helper.tpl  entry for  objstore_uri: 'mongodb://$release_name-centralledger-obj:27017/mlos'  if so also need to do this for admin-api-svc values file too ? 
- please ignore package.sh (I will remove this later, it is simply a time-saving tool for me)

### CODE REVIEW

**Comment by elnyry-sam-k** [comment]
hi @tdaly61 can you please sign the CLA ?

**Comment by tdaly61** [comment]
hi @elnyry-sam-k I have signed the CLA as tdaly61, seems like my .git/config is wrong on my laptop , I will see what I can do to fix and link.

**Comment by lewisdaly** [comment]
> hi @elnyry-sam-k I have signed the CLA as tdaly61, seems like my .git/config is wrong on my laptop , I will see what I can do to fix and link.

You may need to update your git config, and then squash the commits locally, and force push that branch to stop the CLA tool from complaining.

File: scripts/update-charts-dep.sh
Comment (lewisdaly) [inline scripts/update-charts-dep.sh:46]: You can remove this line, and line 48

File: scripts/package.sh
Comment (lewisdaly) [inline scripts/package.sh:40]: I think we will want to remove these changes?

File: scripts/update-charts-dep.sh
Comment (tdaly61) [reply scripts/update-charts-dep.sh]: done

File: mojaloop/chart-admin/values.yaml
Comment (mdebarros) [inline mojaloop/chart-admin/values.yaml:205]: Pls remove, this was probably left by me. Sorry about that. :D

File: mojaloop/chart-admin/values.yaml
Comment (mdebarros) [inline mojaloop/chart-admin/values.yaml:242]: Please remove.

File: mojaloop/account-lookup-service/values.yaml
Comment (mdebarros) [inline mojaloop/account-lookup-service/values.yaml:17]: Please remove.

File: mojaloop/chart-admin/values.yaml
Comment (mdebarros) [inline mojaloop/chart-admin/values.yaml:19]: Please remove.

File: mojaloop/account-lookup-service/values.yaml
Comment (mdebarros) [inline mojaloop/account-lookup-service/values.yaml:840]: Please remove.

File: mojaloop/account-lookup-service/values.yaml
Comment (mdebarros) [inline mojaloop/account-lookup-service/values.yaml:934]: Take note that this has changed in v13.1.0 which applies to all ALS services, ref: https://github.com/mojaloop/helm/blob/v13.1.0/.changelog/release-v13.1.0.md#8-upgrade-notes, see section 8.4 for more details.

I don't know if you want to update here, and align to the match the latest updates, or leave that for a separate story?

These changes would also have to be made to ml-api-adapter....so I am leaning to a new story. 

Let me know what you think.

File: mojaloop/chart-admin/templates/deployment.yaml
Comment (mdebarros) [inline mojaloop/chart-admin/templates/deployment.yaml:144]: Take note that this has changed in v13.1.0 which applies to all ALS services, ref: https://github.com/mojaloop/helm/blob/v13.1.0/.changelog/release-v13.1.0.md#8-upgrade-notes, see section 8.4 for more details.

I don't know if you want to update here, and align to the match the latest updates, or leave that for a separate story?

These changes would also have to be made to ml-api-adapter....so I am leaning to a new story. 

Let me know what you think.

File: mojaloop/chart-admin/values.yaml
Comment (mdebarros) [inline mojaloop/chart-admin/values.yaml:336]: Take note that this has changed in v13.1.0 which applies to all ALS services, ref: https://github.com/mojaloop/helm/blob/v13.1.0/.changelog/release-v13.1.0.md#8-upgrade-notes, see section 8.4 for more details.

I don't know if you want to update here, and align to the match the latest updates, or leave that for a separate story?

These changes would also have to be made to ml-api-adapter....so I am leaning to a new story. 

Let me know what you think.

File: mojaloop/chart-service/templates/deployment.yaml
Comment (mdebarros) [inline mojaloop/chart-service/templates/deployment.yaml:145]: Take note that this has changed in v13.1.0 which applies to all ALS services, ref: https://github.com/mojaloop/helm/blob/v13.1.0/.changelog/release-v13.1.0.md#8-upgrade-notes, see section 8.4 for more details.

I don't know if you want to update here, and align to the match the latest updates, or leave that for a separate story?

These changes would also have to be made to ml-api-adapter....so I am leaning to a new story. 

Let me know what you think.

File: mojaloop/chart-service/values.yaml
Comment (mdebarros) [inline mojaloop/chart-service/values.yaml:150]: Please remove.

File: mojaloop/chart-service/values.yaml
Comment (mdebarros) [inline mojaloop/chart-service/values.yaml:20]: Please remove.

File: mojaloop/chart-service/values.yaml
Comment (mdebarros) [inline mojaloop/chart-service/values.yaml:221]: Please remove.

File: mojaloop/example-backend/values.yaml
Comment (mdebarros) [inline mojaloop/example-backend/values.yaml:2695]: See above.

**Comment by mdebarros** [comment]
Hey @tdaly61 , here is `my response` to some of your points of attention...

> Can I please get your eyes on this an in particular with attention to the following:

* parent HELM chart account-lookup-service is populated via the ./scripts/populate_values.py script   this script copies the longest globals section from the subcharts and then copies the entire subcharts    values.yaml to the parent chart (see account-lookup-service/values.yaml ) `<-- This looks amazing! Great job on this, I will definitely give this a try! Perhaps you could show me a quick demo in tomorrow's discussion?`
* Please check the subcharts values.yaml files carefully and provide feedback `<-- had a review and it looks pretty much ok, just some minor comments. I will give it a 2nd pass tomorrow before our meeting.`
* Please inspect closely the subcharts templates/deployment.yaml **<-- had a review and it looks pretty much ok, just some minor comments. I will give it a 2nd pass tomorrow before our meeting.**
* what in the heck to set pod and node affinities to by default in the values.yaml files ? `<-- In general we would leave these open, and not provide any defaults as they are normally deployment specific.`
* Please note the plan is to update the chart naming to better reflect the reference architecture after most of the updates are done for v14 `<-- Noted`
* should  we create an  example_backend _helper.tpl  entry for  objstore_uri: 'mongodb://$release_name-centralledger-obj:27017/mlos'  if so also need to do this for admin-api-svc values file too ?
* please ignore package.sh (I will remove this later, it is simply a time-saving tool for me) `<-- It should not be required as I believe that this is only needed for the bulk-api-adapter.`

**Comment by tdaly61** [comment]
@mdebarros , thanks for the feedback , most useful and ideally tonight when we meet I am hoping to fix anything remaining and be able to integrate.  

In reply to your questions 

V14 advantages i.e. the good:   (largely what we already discussed/planned)
- separation of the backend services definitely is key
- the extended use of the common/templates and _helpers is good
- testing over multiple K8s versions (ingress / network version etc) provides resilience
- the increased testing/automation is also key e.g. Lewis's circle-ci testing , hopefully the tool to populate top level charts proves generally useful
The bad: 
- values files are still awful / huge / awkward.
- mysql is still "almost" hardcoded in the charts, we could make this less so , whilst still providing a good example using mysql.
-  default passwords really are a bad idea, I think we should remove and provide a small tool to allow user to easily generate / populate values.yaml at deploy-time.If nothing else this looks pretty bad from a security/maturity perspective even if we are not deployers 
- Lewis brought up a good point, it would be good to have all properties/values come from the config files/mounts and a small javascript util I think could do that (@lewisdaly or I should probably file an rfe for that , if one does not already exist)

File: mojaloop/account-lookup-service/Chart.yaml
Comment (tdaly61) [reply mojaloop/account-lookup-service/Chart.yaml:4]: to 13.0.0 ?

File: mojaloop/account-lookup-service/values.yaml
Comment (tdaly61) [reply mojaloop/account-lookup-service/values.yaml:151]: Ok but this needs to happen in the subcharts , remembering that this values.yaml is generated from the populate_values.py tool.

File: mojaloop/admin-api-svc/values.yaml
Comment (tdaly61) [reply mojaloop/admin-api-svc/values.yaml:204]: yep

File: mojaloop/chart-admin/Chart.yaml
Comment (tdaly61) [reply mojaloop/chart-admin/Chart.yaml:4]: again to 13.0.0 ?

File: mojaloop/account-lookup-service/values.yaml
Comment (tdaly61) [reply mojaloop/account-lookup-service/values.yaml:934]: yeah I figured the update to 13.1.0 should be separate as this task i.e. v14-v14  is complex enough trying to do the 13.1 updates at the same time .... seems like a bad idea.  So yeah strongly of the opinion that we should have a separate story.

File: mojaloop/account-lookup-service/values.yaml
Comment (tdaly61) [reply mojaloop/account-lookup-service/values.yaml:316]: as above, separate story.

File: mojaloop/chart-admin/templates/deployment.yaml
Comment (tdaly61) [reply mojaloop/chart-admin/templates/deployment.yaml:144]: separate story

File: mojaloop/chart-admin/values.yaml
Comment (tdaly61) [reply mojaloop/chart-admin/values.yaml:336]: separate story

File: scripts/perl-1-liners.txt
Comment (tdaly61) [reply scripts/perl-1-liners.txt:1]: yeah maybe part of the readme.md ... I will do that and remove this file.

File: scripts/populate_values.py
Comment (tdaly61) [reply scripts/populate_values.py:1]: yeah generally the .py is used to show it is python not bash or sh

File: scripts/populate_values.py
Comment (tdaly61) [reply scripts/populate_values.py:80]: ./scripts/populate_values.py 
usage: populate_values.py [-h] -p PARENTCHART
populate_values.py: error: the following arguments are required: -p/--parentchart

I have added doc to the scripts/readme.md to better document , but essentially it has an ok-ish usage clause and instructions for the single parameter entry.

**Comment by millerabel** [comment]
I have to agree with …
“default passwords really are a bad idea”

Please remove them.

This is an excellent idea…
“…provide a small tool to allow user to easily generate / populate values.yaml at deploy-time.”
We should provide bogus placeholders for passwords that don’t work without config;
Provide a sysgen utility to (as part of its function) generate the required secrets and config values, with ability to use a test seed for repeatability (for debugging); it might over time provide other configuration;

On Mar 15, 2022, at 5:38 PM, Tom Daly ***@***.***> wrote:

﻿
@mdebarros , thanks for the feedback , most useful and ideally tonight when we meet I am hoping to fix anything remaining and be able to integrate.

In reply to your questions

V14 advantages i.e. the good: (largely what we already discussed/planned)

separation of the backend services definitely is key
the extended use of the common/templates and _helpers is good
testing over multiple K8s versions (ingress / network version etc) provides resilience
the increased testing/automation is also key e.g. Lewis's circle-ci testing , hopefully the tool to populate top level charts proves generally useful
The bad:
values files are still awful / huge / awkward.
mysql is still "almost" hardcoded in the charts, we could make this less so , whilst still providing a good example using mysql.
default passwords really are a bad idea, I think we should remove and provide a small tool to allow user to easily generate / populate values.yaml at deploy-time.If nothing else this looks pretty bad from a security/maturity perspective even if we are not deployers
Lewis brought up a good point, it would be good to have all properties/values come from the config files/mounts and a small javascript util I think could do that ***@***.*** or I should probably file an rfe for that , if one does not already exist)
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: scripts/lint-charts.sh
Comment (tdaly61) [reply scripts/lint-charts.sh:25]: yeah I think it was for my testing , I guess I checked in the changes inadvertently , will try undoing this (will test/refresh my git skills  :-)  )

File: mojaloop/example-backend/values.yaml
Comment (tdaly61) [reply mojaloop/example-backend/values.yaml:2387]: To me yes, these should be disabled and then either again a tool to turn on or some simple instructions as to how to enable all the bizops functions.

Comment (tdaly61) [reply mojaloop/example-backend/values.yaml:2387]: further to this, they should be off for development and we could make a final call at the release of V14 as to what is on and off precisely

File: scripts/lint-charts.sh
Comment (tdaly61) [reply scripts/lint-charts.sh:25]: I have updated this and it is not a) same as master branch except b) for the updated account-look-up charts.

File: mojaloop/chart-service/templates/deployment.yaml
Comment (tdaly61) [reply mojaloop/chart-service/templates/deployment.yaml:145]: as above , separate story.

**Comment by tdaly61** [comment]
@millerabel @elnyry-sam-k : so I just filed :  remove default passwords for databases for V14 of ML Helm charts #2720 .  Sam I made it part of the v14 release if this is ok ?

File: mojaloop/account-lookup-service/values.yaml
Comment (tdaly61) [reply mojaloop/account-lookup-service/values.yaml:151]: So ...why is this in the old charts ? 
Is there a way (docs) somewhere I can tell definitively what is needed for each components (other than reading the src, which I have been doing a bit of)

**Comment by tdaly61** [comment]
hi @mdebarros ; so I have the circle-ci tests passing, which means that the new charts and the already converted V14 charts in the charts repo are all "sound".  The issue now is that I have conflicts that prevent a merge and I have a couple of choices, merge offline once you approve OR resolve the merges by pushing a commit that resolves them BUT this will break the circle-ci tests. So I think the best approach from a git/history/doc perspective is if I 1) resolve the merge conflicts by pushing a commit that resolves them BUT breaks the pipeline and then 2) do another small PR that resolves the pipeline issues and removes the Chart.lock that is one of the merge conflicts BUT which per our conversation last night should  not be in the repo.  ...this is what I am gonna do .. then I will ping you in slack and hopefully we can get this part done in the next couple of hrs.

**Review by mdebarros — APPROVED** [review]
+1

**Review by mdebarros — CHANGES_REQUESTED** [review]
@tdaly61 let me know when you are available for a chat about the tests not passing.

**Comment by elnyry-sam-k** [comment]
> @millerabel @elnyry-sam-k : so I just filed : remove default passwords for databases for V14 of ML Helm charts #2720 . Sam I made it part of the v14 release if this is ok ?

Thanks @tdaly61 .. It would be great to have this for v14

**Pull request merged**
